### PR TITLE
ansible-galaxy - add support for submodules when installing collection

### PR DESCRIPTION
**Pull request by danielino**
##### SUMMARY

if a collection has git submodules eg. roles/<submodule1>, ... , install from source code does not initialize git submodules.

this patch check for `<collection_tmp_path>/.gitmodules` presence and then execute `git submodule update --init --recursive`

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

ansible-galaxy collection install git+https://path/to/repo

##### ADDITIONAL INFORMATION

this is an example execution with a collection that contains submodules

```paste below
(.venv) ➜  ansible git:(devel) ✗ bin/ansible-galaxy collection install 'git+https://***gitlab***/******/ansible-collections/abc.xyz.git,master' --force
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under development. This is a rapidly changing
source of code and can become unstable at any point.
Cloning into '/Users/xyz/.ansible/tmp/ansible-local-8633mqv7xdfr/tmpk1qitvel/abc.xyz'...
remote: Enumerating objects: 102, done.
remote: Counting objects: 100% (102/102), done.
remote: Compressing objects: 100% (67/67), done.
remote: Total 778 (delta 46), reused 79 (delta 33), pack-reused 676
Receiving objects: 100% (778/778), 123.61 KiB | 126.00 KiB/s, done.
Resolving deltas: 100% (416/416), done.
Already on 'master'
Your branch is up to date with 'origin/master'.

Submodule 'roles/submodule3' (https://***/****/ansible/ansible-roles/***.git) registered for path 'roles/submodule3'
Submodule path 'roles/submodule3': checked out 'b72e41408991c5ae709f83e67f44e9d4b01e7aeb'
Starting galaxy collection install process
Process install dependency map
Starting collection install process
Installing 'abc.xyz:1.0.3' to '/Users/xyz/.ansible/collections/ansible_collections/abc/xyz'
Created collection for abc.xyz:1.0.3 at /Users/xyz/.ansible/collections/ansible_collections/abc/xyz
abc.xyz:1.0.3 was installed successfully
(.venv) ➜  ansible git:(devel) ✗ 

```

### CODE REVIEW

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py [hunk 1]
```diff
@@ -453,6 +453,18 @@ def _extract_collection_from_git(repo_url, coll_ver, b_path):
...
+        git_submodule_cmd = git_executable, 'submodule', 'update', '--init', '--recursive'
+        try:
+            subprocess.check_call(git_submodule_cmd, cwd=b_checkout_path)
+        except subprocess.CalledProcessError as proc_err:
+            raise_from(
+                AnsibleError(
+                    'Failed to download submodules'
+                ),
+                proc_err,
+            )
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py]: Controller-side code is already py3-only so there shouldn't be any need for the six helper
```suggestion
            raise AnsibleError('Failed to download submodules') from proc_err

```

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py [hunk 2]
```diff
@@ -453,6 +453,18 @@ def _extract_collection_from_git(repo_url, coll_ver, b_path):
             proc_err,
         )
 
+    if pathlib.Path(str(b_checkout_path) + "/.gitmodules").is_file():
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py]: There's no point in using pathlib if you don't actually use it:
```suggestion
    if (pathlib.Path(b_checkout_path) / '.gitmodules').is_file():
```

Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py]: Also, I don't think that doing this unconditionally would be a great idea. If there's gitmodules entries pointing at private or password-protected Git remotes, this would fail. So it'd probably be reasonable to make it depend on a CLI option:
```suggestion
    has_submodules = (pathlib.Path(b_checkout_path) / '.gitmodules').is_file()
    with_submodules_cli_request = ...
    if with_submodules_cli_request and has_submodules:
```

Comment (danielino) [reply lib/ansible/galaxy/collection/concrete_artifact_manager.py]:
```python
    has_submodules = (pathlib.Path(b_checkout_path.decode()) / '.gitmodules').is_file()
    with_submodule_cli_request = context.CLIARGS['init_submodules']
    if with_submodule_cli_request and has_submodules:
```

init_submodules is a boolean flag:

```
(.venv) ➜  ansible git:(devel) ✗ bin/ansible-galaxy collection install --help 2>/dev/null| grep init-submodules
                                         [--init-submodules]
  --init-submodules     Init collection submodules
```

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py [hunk 3]
```diff
@@ -452,6 +454,14 @@ def _extract_collection_from_git(repo_url, coll_ver, b_path):
             ),
             proc_err,
         )
+    has_submodules = (pathlib.Path(b_checkout_path.decode()) / '.gitmodules').is_file()
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py]:
```suggestion

    has_submodules = (pathlib.Path(b_checkout_path.decode()) / '.gitmodules').is_file()
```

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py [hunk 4]
```diff
@@ -453,6 +455,15 @@ def _extract_collection_from_git(repo_url, coll_ver, b_path):
...
         )
 
+    has_submodules = (pathlib.Path(b_checkout_path.decode()) / '.gitmodules').is_file()
+    with_submodule_cli_request = context.CLIARGS['init_submodules']
+    if with_submodule_cli_request and has_submodules:
+        git_submodule_cmd = git_executable, 'submodule', 'update', '--init', '--recursive'
+        try:
+            subprocess.check_call(git_submodule_cmd, cwd=b_checkout_path)
+        except subprocess.CalledProcessError as proc_err:
+            raise AnsibleError('Failed to download submodules') from proc_err
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py]:
```suggestion
            raise AnsibleError('Failed to download Git submodules') from proc_err
```

File: lib/ansible/cli/galaxy.py [hunk 5]
```diff
@@ -498,6 +498,8 @@ def add_install_options(self, parser, parents=None):
             install_parser.add_argument('--ignore-signature-status-code', dest='ignore_gpg_errors', type=str, action='append',
                                         help=ignore_gpg_status_help, default=C.GALAXY_IGNORE_INVALID_SIGNATURE_STATUS_CODES,
                                         choices=list(GPG_ERROR_MAP.keys()))
+            install_parser.add_argument('--init-submodules', dest='init_submodules', action='store_true',
+                                        default=False, help='Init collection submodules')
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py]: This should probably have a mention of Git in the option name and explanation that it's only used when installing from Git SCM and doesn't do anything otherwise.

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py [hunk 6]
```diff
@@ -453,6 +455,15 @@ def _extract_collection_from_git(repo_url, coll_ver, b_path):
             proc_err,
         )
 
+    has_submodules = (pathlib.Path(b_checkout_path.decode()) / '.gitmodules').is_file()
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py]: `.decode()` isn't always present because this module was written at time when it still needed to support both py2 and py3. We have a set of helpers `to_native()`, `to_bytes()` and `to_text()` for dealing with this sort of conversions. It is customary to use them instead as a convention.
```suggestion
    has_submodules = (pathlib.Path(to_native(b_checkout_path)) / '.gitmodules').is_file()
```

**Comment by webknjaz** [comment]
@danielino note that passing CI is not enough to get this merged. This PR will have to go through the triage process where multiple core team members will assess whether it's desired.

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py [hunk 7]
```diff
@@ -453,6 +455,15 @@ def _extract_collection_from_git(repo_url, coll_ver, b_path):
             proc_err,
         )
 
+    has_submodules = (pathlib.Path(to_native(b_checkout_path)) / '.gitmodules').is_file()
+    with_submodule_cli_request = context.CLIARGS['git_init_submodules']
```
Comment (webknjaz) [inline lib/ansible/galaxy/collection/concrete_artifact_manager.py:459]: This value should be passed through the cascade of calls, not accessed directly. It's evident in the failing unit tests how this causes abstraction leaks.

File: lib/ansible/cli/galaxy.py [hunk 8]
```diff
@@ -498,6 +498,8 @@ def add_install_options(self, parser, parents=None):
             install_parser.add_argument('--ignore-signature-status-code', dest='ignore_gpg_errors', type=str, action='append',
                                         help=ignore_gpg_status_help, default=C.GALAXY_IGNORE_INVALID_SIGNATURE_STATUS_CODES,
                                         choices=list(GPG_ERROR_MAP.keys()))
+            install_parser.add_argument('--git-init-submodules', dest='git_init_submodules', action='store_true',
```
Comment (webknjaz) [inline lib/ansible/cli/galaxy.py:501]: This is only added to the `install` subcomand meaning that in other cases there would be access of a non-existing undefined CLI option which is visible in the CI failures.

File: lib/ansible/galaxy/collection/concrete_artifact_manager.py (hunk 7 above)
Comment (danielino) [reply lib/ansible/galaxy/collection/concrete_artifact_manager.py:459]: it's complicated. i think it's better to use context.CLIARGS.get(name, default) as it's done in https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/role.py#L345

this is the call sequence:

image

 what do you think about it?

Comment (webknjaz) [reply lib/ansible/galaxy/collection/concrete_artifact_manager.py:459]: I think it's unacceptable to do this. I wrote this code as well as most things under `lib/ansible/galaxy/dependency_resolution/` and `lib/ansible/galaxy/collection/` with a clear intent to keep it as self-contained as possible. This is why, if you look at the code in the module, it doesn't access magical global contexts.

`lib/ansible/galaxy/role.py` has existed like this historically, which is the only reason it still has these bits of code smell in it. Besides, it's not a fair comparison since you're pointing at a module in a different importable package, on a different abstraction layer. There's no such examples under `lib/ansible/galaxy/collection/*` where you actually do modifications.

OTOH, if you look closer at https://github.com/danielino/ansible/blob/devel/lib/ansible/cli/galaxy.py#L1017-L1021, you'll see that all CLI arguments are inspected and translated there, in that module. And only things necessary for the function from `lib/ansible/galaxy/collection/` are actually passed there, in the correct form.

Also, this value wouldn't need to be passed to every method call in that stack but probably just added to the initializer of `ConcreteArtifactsManager` and stored as an instance attribute and then, accessed via `self` later following the structure that is already in place. Adding an extra approach that doesn't follow literally everything else would make the code more entangled, hard to follow, unmaintainable. We don't want that.

File: lib/ansible/cli/galaxy.py (hunk 8 above)
Comment (s-hertel) [reply lib/ansible/cli/galaxy.py:501]: It can even cause a KeyError when installing collections, because `ansible-galaxy install` will install roles and collections at the same time but doesn't allow collection-specific cli options. We use `context.CLIARGS.get()` for that issue in a few places. Here's an example: https://github.com/s-hertel/ansible/commit/0bfec37608e0bf9af2911f25e426b9d9175a051b.

**Comment by s-hertel** [comment]
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute.

We discussed this in triage today. Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational.
Our reasons for this are:

   * We don't want to encourage the use of submodules in collections. Content of a role is usually highly dependent on the collection it's in, and supporting submodules may mislead people into treating them as generic project dependencies.

Installing collections from git repositories is a convenient developer shortcut, but you can create a playbook to do the same thing with several tasks.

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!
